### ISSUE TITLE
using <ignore> for directory in rootcheck results in link count mismatch for parent directory

### ISSUE THREAD

**Issue by tkutergin** [wazuh/wazuh#4909]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 3.11.3-rev | wazuh-core | Agent | Packages | All |

When <ignore> is used in <rootcheck> section to exclude part of tree from rootcheck, this results in link mismatch for parent directory, for example, using

&lt;ignore&gt;/var/lib/docker&lt;/ignore&gt;

Results in following:
Files hidden inside directory '/var/lib'. Link count does not match number of files (46,45).
Numbers may vary.

This seems to be caused by the following code in 
static int read_sys_dir(const char *dir_name, int do_read)

        /* Ignore the /proc and /sys filesystems */
        if (check_ignore(f_name) || !strcmp(f_name, "/proc") || !strcmp(f_name, "/sys")) {
            continue;
        }

where entry_count is not increased

**Comment by antoniomanuelfr** [comment]
Hi! @tkutergin. 
Thanks for reporting this, I' m going to check this issue. 
Regards.

**Comment by vikman90** [comment]
Possible related issue: #4065

**Comment by tkutergin** [comment]
Probably correct way is to move

        /* Ignore the /proc and /sys filesystems */
        if (check_ignore(f_name) || !strcmp(f_name, "/proc") || !strcmp(f_name, "/sys")) {
            continue;
        }

after

        /* Check if file is a directory */
        if (lstat(f_name, &statbuf_local) == 0) {
            /* On all the systems except Darwin, the
             * link count is only increased on directories
             */
#ifndef Darwin
            if (S_ISDIR(statbuf_local.st_mode))
#else
            if (S_ISDIR(statbuf_local.st_mode) ||
                    S_ISREG(statbuf_local.st_mode)
            /* No S_ISLNK on Windows */
#ifndef WIN32
                    || S_ISLNK(statbuf_local.st_mode)
#endif
                    )
#endif
            {
                entry_count++;
            }
        }

Since this code increases entry_count for directories. At least at first glance this removes false positives.

**Comment by antoniomanuelfr** [comment]
Hi @tkutergin.
First of all, sorry for the late response. 
I have been able to replicate your issue and you are right. I have made the following: 
1. Add the ignore option to syscheck `<ignore>/etc/dummy</ignore>`
2. Create the folder `/etc/dummy_folder`
3. Create some files inside. 
4. Restart wazuh. 
5. The following alert is generated: 
```
** Alert 1587109472.9985: - ossec,rootcheck,gdpr_IV_35.7.d,
2020 Apr 17 07:44:32 ubuntumanager->rootcheck
Rule: 510 (level 7) -> 'Host-based anomaly detection event (rootcheck).'
Files hidden inside directory '/etc'. Link count does not match number of files (96,97).
title: Files hidden inside directory '/etc'.
```
This is still happening in 3.12 version.
Thanks for reporting this issue. We will try to fix it as soon as possible. 
Regards. 
Antonio

**Comment by antoniomanuelfr** [comment]
Hi again @tkutergin. 
We have this [pull request](https://github.com/wazuh/wazuh/pull/4603) that will solve the problem and we hope that it will be merged as soon as our workflow allow us. 
Regards

**Comment by vikman90** [comment]
@tkutergin @antoniomanuelfr this issue has been fixed by #4603, which will be available in the next release 3.0.0.

Thank you for reporting this bug.
Best regards.

**Comment by dbray925** [comment]
This issue seems to be happening again on current version 4.10.0.

    <ignore>/var/lib/clickhouse</ignore>
    <ignore>/var/lib/docker/volumes</ignore>

But yet:
Rule: 510 fired (level 7) -> "Host-based anomaly detection event (rootcheck)."
Portion of the log(s):

Files hidden inside directory '/var/lib/clickhouse/store/a34/a343b323-bbc4-49b5-8ac6-47e9e424c5d6'. Link count does not match number of files (782,781).

Should I open a new issue?
